accepted/future-releases/generalized-typedef-2018/feature-specification.md

```diff
@@ -0,0 +1,149 @@
+`Map` is changed to `SnackBar`, but such mistakes could easily happen
+during development. It may not even be easy to detect exactly where the
+mistake occurred, in cases where the erroneous type is used in some
+declaration, and expressions whose types depend on that declaration
+unexpectedly get flagged as compile-time errors, or the IDE completions
+in such expressions fail to list some of the expected choices.
+
+This document describes how type aliases are generalized in Dart 2.2
+to make such verbose types more concise and consistent.
+
+
+## Feature Specification
+
+The precise definition of the changes is given in the language
+specification, with a proposed form in
+[this CL](https://dart-review.googlesource.com/c/sdk/+/81414).
+The following sections summarize the changes.
+
+
+### Syntax
+
+The grammar is modified as follows in order to support this feature:
+
+```
+typeAlias ::=
+  metadata 'typedef' typeAliasBody |
+  metadata 'typedef' identifier typeParameters? '=' type ';' // CHANGED
+```
+
+*The modification is that a type alias declaration of the form that uses
+`=` can define the type alias to denote any type, not just a function
+type.*
+
+
+### Static Analysis 
+
+There is no change in the treatment of existing syntax, so we describe only
+the treatment of syntactic forms that can be new when this feature is added.
+
+The effect of a non-generic type alias declaration of the form that uses
+`=` with name `F` is to bind the name `F` in the library scope of the
+enclosing library to the type denoted by the right hand side of `=` in that
+declaration.
+
+*This is not new, but it applies to a larger set of situations now that the
+right hand side can be any type. Let us call that type on the right hand
+side `T`. This means that `F` can be used as a type annotation, and the
+entity which is given type `F` is considered to have type `T`, with the
+same meaning as in the declaration of `F` (even if some declaration in the
+current library gives a new meaning to an identifier in `T`).*
+
+The effect of a generic type alias declaration of the form
+
+```dart
+typedef F<X1 extends B1 .. Xk extends Bk> = T;
+```
+
+where metadata is omitted but may be present, is to bind the name `F` in
+the library scope of the enclosing library to a mapping from type argument
+lists to types, such that a parameterized type of the form `F<T1..Tk>` is
+an application of that mapping that denotes the type `[T1/X1 .. Tk/Xk]T`.
+
+Let _F_ be a generic type alias of the form that uses `=` with type
+parameter declarations
+_X<sub>1</sub> extends B<sub>1</sub> .. X<sub>k</sub> extends B<sub>k</sub>_.
+It is a compile-time error unless satisfaction of the declared bounds
+_B<sub>1</sub> .. B<sub>k</sub>_ implies that the right hand side is
+regular-bounded, and all types that occur as subterms of the right hand
+side are well-bounded.
+Any self reference in a type alias, either directly or recursively via another
```

Comment:
As mentioned in the spec CL, such self-references must be absent after instantiate-to-bounds on the RHS. Effectively, I guess we will "evaluate" the RHS as a type expression and complain if it ever needs to refer to the LHS.